File: 15-design-patterns-2/src/main/python/singleton/database_connection.py
from typing import Optional


class DatabaseConnection:
    _instance: Optional["DatabaseConnection"] = None

    def __new__(cls) -> "DatabaseConnection":
        if cls._instance is None:
            cls._instance = super(DatabaseConnection, cls).__new__(cls)
            cls._instance.driver = "org.postgresql.Driver"
            cls._instance.url = \
                "jdbc:postgresql://localhost:5432/postgres"
        return cls._instance


if __name__ == "__main__":
    db1 = DatabaseConnection()
    db2 = DatabaseConnection()
    print(db1 is db2) # True

File: 15-design-patterns-2/src/main/scala/singleton/DatabaseConnection.scala
package cl.uchile.dcc
package singleton

class DatabaseConnection private (val driver: String, val url: String)
// Controla la instancia única desde el objeto compañero
object DatabaseConnection:
  private var instance: Option[DatabaseConnection] = None

  def getInstance: DatabaseConnection =
    if instance.isEmpty then
      instance = Some(
          new DatabaseConnection("org.postgresql.Driver",
                                 "jdbc:postgresql://localhost:5432/test"
          )
      )
    instance.get
